File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/OptionFeature.cpp
// OptionFeature.cpp 
#include "OptionFeature.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

OptionFeature::OptionFeature(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //fxFeatureNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fxFeatureNode = xmlNode->FirstChildElement("fxFeature");

   if(fxFeatureNode){fxFeatureIsNull_ = false;}
   else{fxFeatureIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fxFeatureNode , address : " << fxFeatureNode << std::endl;
   #endif
   if(fxFeatureNode)
   {
      if(fxFeatureNode->Attribute("href") || fxFeatureNode->Attribute("id"))
      {
          if(fxFeatureNode->Attribute("id"))
          {
             fxFeatureIDRef_ = fxFeatureNode->Attribute("id");
             fxFeature_ = boost::shared_ptr<FxFeature>(new FxFeature(fxFeatureNode));
             fxFeature_->setID(fxFeatureIDRef_);
             IDManager::instance().setID(fxFeatureIDRef_,fxFeature_);
          }
          else if(fxFeatureNode->Attribute("href")) { fxFeatureIDRef_ = fxFeatureNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fxFeature_ = boost::shared_ptr<FxFeature>(new FxFeature(fxFeatureNode));}
   }

   //strategyFeatureNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* strategyFeatureNode = xmlNode->FirstChildElement("strategyFeature");

   if(strategyFeatureNode){strategyFeatureIsNull_ = false;}
   else{strategyFeatureIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- strategyFeatureNode , address : " << strategyFeatureNode << std::endl;
   #endif
   if(strategyFeatureNode)
   {
      if(strategyFeatureNode->Attribute("href") || strategyFeatureNode->Attribute("id"))
      {
          if(strategyFeatureNode->Attribute("id"))
          {
             strategyFeatureIDRef_ = strategyFeatureNode->Attribute("id");
             strategyFeature_ = boost::shared_ptr<StrategyFeature>(new StrategyFeature(strategyFeatureNode));
             strategyFeature_->setID(strategyFeatureIDRef_);
             IDManager::instance().setID(strategyFeatureIDRef_,strategyFeature_);
          }
          else if(strategyFeatureNode->Attribute("href")) { strategyFeatureIDRef_ = strategyFeatureNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { strategyFeature_ = boost::shared_ptr<StrategyFeature>(new StrategyFeature(strategyFeatureNode));}
   }

   //asianNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* asianNode = xmlNode->FirstChildElement("asian");

   if(asianNode){asianIsNull_ = false;}
   else{asianIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- asianNode , address : " << asianNode << std::endl;
   #endif
   if(asianNode)
   {
      if(asianNode->Attribute("href") || asianNode->Attribute("id"))
      {
          if(asianNode->Attribute("id"))
          {
             asianIDRef_ = asianNode->Attribute("id");
             asian_ = boost::shared_ptr<Asian>(new Asian(asianNode));
             asian_->setID(asianIDRef_);
             IDManager::instance().setID(asianIDRef_,asian_);
          }
          else if(asianNode->Attribute("href")) { asianIDRef_ = asianNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { asian_ = boost::shared_ptr<Asian>(new Asian(asianNode));}
   }

   //barrierNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* barrierNode = xmlNode->FirstChildElement("barrier");

   if(barrierNode){barrierIsNull_ = false;}
   else{barrierIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- barrierNode , address : " << barrierNode << std::endl;
   #endif
   if(barrierNode)
   {
      if(barrierNode->Attribute("href") || barrierNode->Attribute("id"))
      {
          if(barrierNode->Attribute("id"))
          {
             barrierIDRef_ = barrierNode->Attribute("id");
             barrier_ = boost::shared_ptr<Barrier>(new Barrier(barrierNode));
             barrier_->setID(barrierIDRef_);
             IDManager::instance().setID(barrierIDRef_,barrier_);
          }
          else if(barrierNode->Attribute("href")) { barrierIDRef_ = barrierNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { barrier_ = boost::shared_ptr<Barrier>(new Barrier(barrierNode));}
   }

   //knockNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* knockNode = xmlNode->FirstChildElement("knock");

   if(knockNode){knockIsNull_ = false;}
   else{knockIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- knockNode , address : " << knockNode << std::endl;
   #endif
   if(knockNode)
   {
      if(knockNode->Attribute("href") || knockNode->Attribute("id"))
      {
          if(knockNode->Attribute("id"))
          {
             knockIDRef_ = knockNode->Attribute("id");
             knock_ = boost::shared_ptr<Knock>(new Knock(knockNode));
             knock_->setID(knockIDRef_);
             IDManager::instance().setID(knockIDRef_,knock_);
          }
          else if(knockNode->Attribute("href")) { knockIDRef_ = knockNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { knock_ = boost::shared_ptr<Knock>(new Knock(knockNode));}
   }

   //passThroughNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* passThroughNode = xmlNode->FirstChildElement("passThrough");

   if(passThroughNode){passThroughIsNull_ = false;}
   else{passThroughIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- passThroughNode , address : " << passThroughNode << std::endl;
   #endif
   if(passThroughNode)
   {
      if(passThroughNode->Attribute("href") || passThroughNode->Attribute("id"))
      {
          if(passThroughNode->Attribute("id"))
          {
             passThroughIDRef_ = passThroughNode->Attribute("id");
             passThrough_ = boost::shared_ptr<PassThrough>(new PassThrough(passThroughNode));
             passThrough_->setID(passThroughIDRef_);
             IDManager::instance().setID(passThroughIDRef_,passThrough_);
          }
          else if(passThroughNode->Attribute("href")) { passThroughIDRef_ = passThroughNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { passThrough_ = boost::shared_ptr<PassThrough>(new PassThrough(passThroughNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<FxFeature> OptionFeature::getFxFeature()
{
   if(!this->fxFeatureIsNull_){
        if(fxFeatureIDRef_ != ""){
             return boost::shared_static_cast<FxFeature>(IDManager::instance().getID(fxFeatureIDRef_));
        }else{
             return this->fxFeature_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<FxFeature>();
   }
}
boost::shared_ptr<StrategyFeature> OptionFeature::getStrategyFeature()
{
   if(!this->strategyFeatureIsNull_){
        if(strategyFeatureIDRef_ != ""){
             return boost::shared_static_cast<StrategyFeature>(IDManager::instance().getID(strategyFeatureIDRef_));
        }else{
             return this->strategyFeature_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<StrategyFeature>();
   }
}
boost::shared_ptr<Asian> OptionFeature::getAsian()
{
   if(!this->asianIsNull_){
        if(asianIDRef_ != ""){
             return boost::shared_static_cast<Asian>(IDManager::instance().getID(asianIDRef_));
        }else{
             return this->asian_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Asian>();
   }
}
boost::shared_ptr<Barrier> OptionFeature::getBarrier()
{
   if(!this->barrierIsNull_){
        if(barrierIDRef_ != ""){
             return boost::shared_static_cast<Barrier>(IDManager::instance().getID(barrierIDRef_));
        }else{
             return this->barrier_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Barrier>();
   }
}
boost::shared_ptr<Knock> OptionFeature::getKnock()
{
   if(!this->knockIsNull_){
        if(knockIDRef_ != ""){
             return boost::shared_static_cast<Knock>(IDManager::instance().getID(knockIDRef_));
        }else{
             return this->knock_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Knock>();
   }
}
boost::shared_ptr<PassThrough> OptionFeature::getPassThrough()
{
   if(!this->passThroughIsNull_){
        if(passThroughIDRef_ != ""){
             return boost::shared_static_cast<PassThrough>(IDManager::instance().getID(passThroughIDRef_));
        }else{
             return this->passThrough_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PassThrough>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/OptionFeature.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class OptionFeature : ISerialized
    {
        public OptionFeature(XmlNode xmlNode)
        {
            XmlNode fxFeatureNode = xmlNode.SelectSingleNode("fxFeature");
            
            if (fxFeatureNode != null)
            {
                if (fxFeatureNode.Attributes["href"] != null || fxFeatureNode.Attributes["id"] != null) 
                {
                    if (fxFeatureNode.Attributes["id"] != null) 
                    {
                        fxFeatureIDRef_ = fxFeatureNode.Attributes["id"].Value;
                        FxFeature ob = new FxFeature(fxFeatureNode);
                        IDManager.SetID(fxFeatureIDRef_, ob);
                    }
                    else if (fxFeatureNode.Attributes["href"] != null)
                    {
                        fxFeatureIDRef_ = fxFeatureNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fxFeature_ = new FxFeature(fxFeatureNode);
                    }
                }
                else
                {
                    fxFeature_ = new FxFeature(fxFeatureNode);
                }
            }
            
        
            XmlNode strategyFeatureNode = xmlNode.SelectSingleNode("strategyFeature");
            
            if (strategyFeatureNode != null)
            {
                if (strategyFeatureNode.Attributes["href"] != null || strategyFeatureNode.Attributes["id"] != null) 
                {
                    if (strategyFeatureNode.Attributes["id"] != null) 
                    {
                        strategyFeatureIDRef_ = strategyFeatureNode.Attributes["id"].Value;
                        StrategyFeature ob = new StrategyFeature(strategyFeatureNode);
                        IDManager.SetID(strategyFeatureIDRef_, ob);
                    }
                    else if (strategyFeatureNode.Attributes["href"] != null)
                    {
                        strategyFeatureIDRef_ = strategyFeatureNode.Attributes["href"].Value;
                    }
                    else
                    {
                        strategyFeature_ = new StrategyFeature(strategyFeatureNode);
                    }
                }
                else
                {
                    strategyFeature_ = new StrategyFeature(strategyFeatureNode);
                }
            }
            
        
            XmlNode asianNode = xmlNode.SelectSingleNode("asian");
            
            if (asianNode != null)
            {
                if (asianNode.Attributes["href"] != null || asianNode.Attributes["id"] != null) 
                {
                    if (asianNode.Attributes["id"] != null) 
                    {
                        asianIDRef_ = asianNode.Attributes["id"].Value;
                        Asian ob = new Asian(asianNode);
                        IDManager.SetID(asianIDRef_, ob);
                    }
                    else if (asianNode.Attributes["href"] != null)
                    {
                        asianIDRef_ = asianNode.Attributes["href"].Value;
                    }
                    else
                    {
                        asian_ = new Asian(asianNode);
                    }
                }
                else
                {
                    asian_ = new Asian(asianNode);
                }
            }
            
        
            XmlNode barrierNode = xmlNode.SelectSingleNode("barrier");
            
            if (barrierNode != null)
            {
                if (barrierNode.Attributes["href"] != null || barrierNode.Attributes["id"] != null) 
                {
                    if (barrierNode.Attributes["id"] != null) 
                    {
                        barrierIDRef_ = barrierNode.Attributes["id"].Value;
                        Barrier ob = new Barrier(barrierNode);
                        IDManager.SetID(barrierIDRef_, ob);
                    }
                    else if (barrierNode.Attributes["href"] != null)
                    {
                        barrierIDRef_ = barrierNode.Attributes["href"].Value;
                    }
                    else
                    {
                        barrier_ = new Barrier(barrierNode);
                    }
                }
                else
                {
                    barrier_ = new Barrier(barrierNode);
                }
            }
            
        
            XmlNode knockNode = xmlNode.SelectSingleNode("knock");
            
            if (knockNode != null)
            {
                if (knockNode.Attributes["href"] != null || knockNode.Attributes["id"] != null) 
                {
                    if (knockNode.Attributes["id"] != null) 
                    {
                        knockIDRef_ = knockNode.Attributes["id"].Value;
                        Knock ob = new Knock(knockNode);
                        IDManager.SetID(knockIDRef_, ob);
                    }
                    else if (knockNode.Attributes["href"] != null)
                    {
                        knockIDRef_ = knockNode.Attributes["href"].Value;
                    }
                    else
                    {
                        knock_ = new Knock(knockNode);
                    }
                }
                else
                {
                    knock_ = new Knock(knockNode);
                }
            }
            
        
            XmlNode passThroughNode = xmlNode.SelectSingleNode("passThrough");
            
            if (passThroughNode != null)
            {
                if (passThroughNode.Attributes["href"] != null || passThroughNode.Attributes["id"] != null) 
                {
                    if (passThroughNode.Attributes["id"] != null) 
                    {
                        passThroughIDRef_ = passThroughNode.Attributes["id"].Value;
                        PassThrough ob = new PassThrough(passThroughNode);
                        IDManager.SetID(passThroughIDRef_, ob);
                    }
                    else if (passThroughNode.Attributes["href"] != null)
                    {
                        passThroughIDRef_ = passThroughNode.Attributes["href"].Value;
                    }
                    else
                    {
                        passThrough_ = new PassThrough(passThroughNode);
                    }
                }
                else
                {
                    passThrough_ = new PassThrough(passThroughNode);
                }
            }
            
        
        }
        
    
        #region FxFeature_
        private FxFeature fxFeature_;
        public FxFeature FxFeature_
        {
            get
            {
                if (this.fxFeature_ != null)
                {
                    return this.fxFeature_; 
                }
                else if (this.fxFeatureIDRef_ != null)
                {
                    fxFeature_ = IDManager.getID(fxFeatureIDRef_) as FxFeature;
                    return this.fxFeature_; 
                }
                else
                {
                      return this.fxFeature_; 
                }
            }
            set
            {
                if (this.fxFeature_ != value)
                {
                    this.fxFeature_ = value;
                }
            }
        }
        #endregion
        
        public string fxFeatureIDRef_ { get; set; }
        #region StrategyFeature_
        private StrategyFeature strategyFeature_;
        public StrategyFeature StrategyFeature_
        {
            get
            {
                if (this.strategyFeature_ != null)
                {
                    return this.strategyFeature_; 
                }
                else if (this.strategyFeatureIDRef_ != null)
                {
                    strategyFeature_ = IDManager.getID(strategyFeatureIDRef_) as StrategyFeature;
                    return this.strategyFeature_; 
                }
                else
                {
                      return this.strategyFeature_; 
                }
            }
            set
            {
                if (this.strategyFeature_ != value)
                {
                    this.strategyFeature_ = value;
                }
            }
        }
        #endregion
        
        public string strategyFeatureIDRef_ { get; set; }
        #region Asian_
        private Asian asian_;
        public Asian Asian_
        {
            get
            {
                if (this.asian_ != null)
                {
                    return this.asian_; 
                }
                else if (this.asianIDRef_ != null)
                {
                    asian_ = IDManager.getID(asianIDRef_) as Asian;
                    return this.asian_; 
                }
                else
                {
                      return this.asian_; 
                }
            }
            set
            {
                if (this.asian_ != value)
                {
                    this.asian_ = value;
                }
            }
        }
        #endregion
        
        public string asianIDRef_ { get; set; }
        #region Barrier_
        private Barrier barrier_;
        public Barrier Barrier_
        {
            get
            {
                if (this.barrier_ != null)
                {
                    return this.barrier_; 
                }
                else if (this.barrierIDRef_ != null)
                {
                    barrier_ = IDManager.getID(barrierIDRef_) as Barrier;
                    return this.barrier_; 
                }
                else
                {
                      return this.barrier_; 
                }
            }
            set
            {
                if (this.barrier_ != value)
                {
                    this.barrier_ = value;
                }
            }
        }
        #endregion
        
        public string barrierIDRef_ { get; set; }
        #region Knock_
        private Knock knock_;
        public Knock Knock_
        {
            get
            {
                if (this.knock_ != null)
                {
                    return this.knock_; 
                }
                else if (this.knockIDRef_ != null)
                {
                    knock_ = IDManager.getID(knockIDRef_) as Knock;
                    return this.knock_; 
                }
                else
                {
                      return this.knock_; 
                }
            }
            set
            {
                if (this.knock_ != value)
                {
                    this.knock_ = value;
                }
            }
        }
        #endregion
        
        public string knockIDRef_ { get; set; }
        #region PassThrough_
        private PassThrough passThrough_;
        public PassThrough PassThrough_
        {
            get
            {
                if (this.passThrough_ != null)
                {
                    return this.passThrough_; 
                }
                else if (this.passThroughIDRef_ != null)
                {
                    passThrough_ = IDManager.getID(passThroughIDRef_) as PassThrough;
                    return this.passThrough_; 
                }
                else
                {
                      return this.passThrough_; 
                }
            }
            set
            {
                if (this.passThrough_ != value)
                {
                    this.passThrough_ = value;
                }
            }
        }
        #endregion
        
        public string passThroughIDRef_ { get; set; }
        
    
        
    
    }
    
}

